grep: fix output duplication when number of matches is greater than MAX_MATCHES

vchimishuk opened the pull request:
When color is enable grep(1) searches for MAX_MATCHES, prints the line and then repeats these steps again untill all matches in the line are found. The issue is that every time grep(1) prints the whole line which leads to the situation when line is duplicated (tripled, quadruple, ...) in stdout.

Bug can be reproduced with the next simple command. It prints line twice when only single line is expected.

$ for i in $(seq 33); do echo -n "foobar"; done | ./grep --color=auto foo

jlduran commented:
> By the way, I tried to run tests in `usr.bin/grep/tests` folder using `kyua test` command on `main` branch and looks like tests are broken or something. Many of them fails with `failed: atf-check failed;` message.

You should be able to run both test suites, the one from FreeBSD you pointed out and the one from NetBSD (`contrib/netbsd-tests/usr.bin/grep`).
Both get installed (usually) under `/usr/tests`, more specifically under `usr.bin/grep`.

<details>

<summary>For example:</summary>

### Tested with Cirrus CI

https://github.com/jlduran/freebsd-src/tree/gh_1442
https://cirrus-ci.com/task/6524118865018880

```
make -C /usr/src/usr.bin/grep -j$(sysctl -n hw.ncpu) all install
kyua test -k /usr/tests/Kyuafile usr.bin/grep/
usr.bin/grep/grep_freebsd_test:gnuext  ->  passed  [0.045s]
usr.bin/grep/grep_freebsd_test:grep_r_implied  ->  passed  [0.019s]
usr.bin/grep/grep_freebsd_test:rgrep  ->  passed  [0.017s]
usr.bin/grep/grep_freebsd_test:zflag  ->  passed  [0.009s]
usr.bin/grep/grep_test:badcontext  ->  passed  [0.029s]
usr.bin/grep/grep_test:basic  ->  passed  [0.018s]
usr.bin/grep/grep_test:begin_end  ->  passed  [0.019s]
usr.bin/grep/grep_test:binary  ->  passed  [0.011s]
usr.bin/grep/grep_test:binary_flags  ->  passed  [0.044s]
usr.bin/grep/grep_test:cflag  ->  passed  [0.026s]
usr.bin/grep/grep_test:color  ->  passed  [0.033s]
usr.bin/grep/grep_test:context  ->  passed  [0.043s]
usr.bin/grep/grep_test:context2  ->  passed  [0.023s]
usr.bin/grep/grep_test:egrep  ->  passed  [0.016s]
usr.bin/grep/grep_test:egrep_empty_invalid  ->  passed  [0.014s]
usr.bin/grep/grep_test:egrep_sanity  ->  passed  [0.032s]
usr.bin/grep/grep_test:emptyfile  ->  passed  [0.015s]
usr.bin/grep/grep_test:escmap  ->  passed  [0.013s]
usr.bin/grep/grep_test:excessive_matches  ->  passed  [0.039s]
usr.bin/grep/grep_test:f_file_empty  ->  passed  [0.010s]
usr.bin/grep/grep_test:fgrep_icase  ->  passed  [0.019s]
usr.bin/grep/grep_test:fgrep_multipattern  ->  passed  [0.016s]
usr.bin/grep/grep_test:fgrep_oflag  ->  passed  [0.044s]
usr.bin/grep/grep_test:fgrep_sanity  ->  passed  [0.017s]
usr.bin/grep/grep_test:file_exp  ->  passed  [0.016s]
usr.bin/grep/grep_test:grep_nomatch_flags  ->  passed  [0.053s]
usr.bin/grep/grep_test:grep_sanity  ->  passed  [0.026s]
usr.bin/grep/grep_test:ignore_case  ->  passed  [0.013s]
usr.bin/grep/grep_test:invert  ->  passed  [0.012s]
usr.bin/grep/grep_test:matchall  ->  passed  [0.025s]
usr.bin/grep/grep_test:mflag  ->  passed  [0.021s]
usr.bin/grep/grep_test:mflag_trail_ctx  ->  passed  [0.020s]
usr.bin/grep/grep_test:mmap  ->  passed  [0.018s]
usr.bin/grep/grep_test:negative  ->  passed  [0.012s]
usr.bin/grep/grep_test:nonexistent  ->  passed  [0.010s]
usr.bin/grep/grep_test:ocolor_metadata  ->  passed  [0.034s]
usr.bin/grep/grep_test:oflag_zerolen  ->  passed  [0.047s]
usr.bin/grep/grep_test:recurse  ->  passed  [0.017s]
usr.bin/grep/grep_test:recurse_symlink  ->  passed  [0.016s]
usr.bin/grep/grep_test:wflag_emptypat  ->  passed  [0.019s]
usr.bin/grep/grep_test:whole_line  ->  passed  [0.015s]
usr.bin/grep/grep_test:word_regexps  ->  passed  [0.018s]
usr.bin/grep/grep_test:wv_combo_break  ->  passed  [0.021s]
usr.bin/grep/grep_test:xflag  ->  passed  [0.012s]
usr.bin/grep/grep_test:xflag_emptypat  ->  passed  [0.045s]
usr.bin/grep/grep_test:xflag_emptypat_plus  ->  passed  [0.020s]
usr.bin/grep/grep_test:zerolen  ->  passed  [0.017s]
usr.bin/grep/grep_test:zgrep  ->  passed  [0.019s]
usr.bin/grep/grep_test:zgrep_combined_flags  ->  expected_failure: known but unsolved zgrep wrapper script regression: atf-check failed; see the output of the test for details  [0.029s]
usr.bin/grep/grep_test:zgrep_eflag  ->  passed  [0.023s]
usr.bin/grep/grep_test:zgrep_empty_eflag  ->  passed  [0.017s]
usr.bin/grep/grep_test:zgrep_fflag  ->  passed  [0.023s]
usr.bin/grep/grep_test:zgrep_long_eflag  ->  passed  [0.015s]
usr.bin/grep/grep_test:zgrep_multiple_eflags  ->  expected_failure: known but unsolved zgrep wrapper script regression: atf-check failed; see the output of the test for details  [0.016s]
usr.bin/grep/grep_test:zgrep_multiple_files  ->  passed  [0.033s]
Results file id is usr_tests.20241003-063631-149783
Results saved to /.kyua/store/results.usr_tests.20241003-063631-149783.db
55/55 passed (0 failed)
```

</details>

vchimishuk commented:
> > By the way, I tried to run tests in `usr.bin/grep/tests` folder using `kyua test` command on `main` branch and looks like tests are broken or something. Many of them fails with `failed: atf-check failed;` message.
> 
> You should be able to run both test suites, the one from FreeBSD you pointed out and the one from NetBSD (`contrib/netbsd-tests/usr.bin/grep`). Both get installed (usually) under `/usr/tests`, more specifically under `usr.bin/grep`.
> 
> For example:

I see now. So, tests are passing, that's a good news.
Thank you.

jlduran commented:
I've added a dumb test:

https://github.com/jlduran/freebsd-src/commit/d2a23d68107f1d8c0d00f883d8529f4d4b0ce6b3

It would be good if a regression test is added as well.

Given the test is under `contrib`, try wrapping it with `# Begin FreeBSD` and `# End FreeBSD` comments

vchimishuk commented:
Thank you, @jlduran. I'll add tests tomorrow. I want to run tests locally to make sure everything is fine.

vchimishuk commented:
@jlduran, I've added a commit with the testcase you proposed. Unfortunately I was not able to run tests locally, looks like I messed up with permissions or something. `make buildworld` takes 7 hours on my machine, so I left it for now. Hope Github CI verification will do the check.
Thank you for helping with the testcase.

jlduran commented on contrib/netbsd-tests/usr.bin/grep/t_grep.sh:
```diff
@@ -426,6 +426,13 @@ color_body()
 
 	atf_check -o file:"$(atf_get_srcdir)/d_color_c.out" \
 	    grep --color=always -f grepfile "$(atf_get_srcdir)/d_color_b.in"
+
```
This empty line is not present in the NetBSD test suite. While aesthetically pleasing, it worsens the diffs when importing the changes from NetBSD.

vchimishuk replied:
Right. Fixed.

jlduran reviewed (commented):
Nice, thank you!
There should be no need to `buildworld`, try:

    make -C /usr/src/usr.bin/grep -j$(sysctl -n hw.ncpu) all install
    kyua test -k /usr/tests/Kyuafile usr.bin/grep/

vchimishuk commented:
> Nice, thank you! There should be no need to `buildworld`, try:
> 
> ```
> make -C /usr/src/usr.bin/grep -j$(sysctl -n hw.ncpu) all install
> kyua test -k /usr/tests/Kyuafile usr.bin/grep/
> ```

Without `buildworld` `/usr/tests/Kyuafile` file was missing for me.

I have the next error by the way. But `/usr/tests/usr.bin/grep/` folder seems to be fine and contains test files.
```
$ sudo kyua test -k /usr/tests/Kyuafile usr.bin/grep/
Results file id is usr_tests.20241007-122554-057863          
Results saved to /root/.kyua/store/results.usr_tests.20241007-122554-057863.db
kyua: W: No test cases matched by the filter 'usr.bin/grep'.
```

jlduran commented:
> > Nice, thank you! There should be no need to `buildworld`, try:
> > ```
> > make -C /usr/src/usr.bin/grep -j$(sysctl -n hw.ncpu) all install
> > kyua test -k /usr/tests/Kyuafile usr.bin/grep/
> > ```
> 
> Without `buildworld` `/usr/tests/Kyuafile` file was missing for me.
> 
> I have the next error by the way. But `/usr/tests/usr.bin/grep/` folder seems to be fine and contains test files.
> 
> ```
> $ sudo kyua test -k /usr/tests/Kyuafile usr.bin/grep/
> Results file id is usr_tests.20241007-122554-057863          
> Results saved to /root/.kyua/store/results.usr_tests.20241007-122554-057863.db
> kyua: W: No test cases matched by the filter 'usr.bin/grep'.
> ```

Try changing into the /usr/tests/usr.bin/grep directory and listing the tests:

```console
$ cd /usr/tests/usr.bin/grep
$ kyua list
grep_freebsd_test:gnuext
grep_freebsd_test:grep_r_implied
...
```

There should be a `Kyuafile`.  If there are no syntax errors with the tests, the list shows up.
You should be able to run the tests:

```console
# kyua test
```

Or debug a specific test:

```console
# kyua debug grep_test:color
Executing command [ grep --color=auto -e .* -e a /usr/tests/usr.bin/grep/d_color_a.in ]
Executing command [ grep --color=auto -f grepfile /usr/tests/usr.bin/grep/d_color_b.in ]
Executing command [ grep --color=always -f grepfile /usr/tests/usr.bin/grep/d_color_b.in ]
Executing command [ grep --color=always foo grepfile ]
grep_test:color  ->  passed
```

vchimishuk commented:
`kyua` works just fine when executing directly from `/usr/tests/usr.bin/grep/`. Thank you!
I need to learn more about this tool. Using it for the first time.

```console
$ kyua debug grep_test:color
Executing command [ grep --color=auto -e .* -e a /usr/tests/usr.bin/grep/d_color_a.in ]
Executing command [ grep --color=auto -f grepfile /usr/tests/usr.bin/grep/d_color_b.in ]
Executing command [ grep --color=always -f grepfile /usr/tests/usr.bin/grep/d_color_b.in ]
Executing command [ grep --color=always foo grepfile ]
grep_test:color  ->  passed
$
```

vchimishuk commented:
[jlduran](https://github.com/jlduran), is there any plans to merge the patch into upstream?

markjdb commented:
@kevans91 would you have any time to review the change?

kevans91 commented on usr.bin/grep/util.c:
```diff
@@ -223,6 +225,9 @@ procmatch_match(struct mprintc *mc, struct parsec *pc)
 			else
 				break;
 		}
+		pc->matchidx = 0;
```
Nit: let's add a blank line above this, please

vchimishuk replied:
Added.

kevans91 reviewed (approved):
LGTM, one nit that we can probably just do pre-commit

bsdimp commented:
The conflicts that were detected were larger than I felt comfortable resolving. This needs to be rebased to proceed.

kevans91 commented:
Sorry, someone had woken up a patch that I had from two or three years ago and I committed that in the interim, which seems to have been the exact problem from here that I'd forgotten about... see 4c9ffb13dd74159bd3ed7e1c4c706dbd15a70df2.  We should just extract the test additions and commit those from here, I think.

kevans91 commented:
Additional contet for their problem is here: https://lists.freebsd.org/archives/freebsd-current/2025-March/007176.html

vchimishuk commented:
> The conflicts that were detected were larger than I felt comfortable resolving. This needs to be rebased to proceed.

@bsdimp, @kevans91, should I updated my PR to resolve merge conflicts or the problem is going to be fixed using different patch?

kevans91 commented:
> > The conflicts that were detected were larger than I felt comfortable resolving. This needs to be rebased to proceed.
> 
> @bsdimp, @kevans91, should I updated my PR to resolve merge conflicts or the problem is going to be fixed using different patch?

Can you confirm your test cases against the latest main (or stable/14), please?  The problem should be fixed already- if it is, then let's pare this down to your test additions and land that.

vchimishuk commented:
> > > The conflicts that were detected were larger than I felt comfortable resolving. This needs to be rebased to proceed.
> > 
> > 
> > @bsdimp, @kevans91, should I updated my PR to resolve merge conflicts or the problem is going to be fixed using different patch?
> 
> Can you confirm your test cases against the latest main (or stable/14), please? The problem should be fixed already- if it is, then let's pare this down to your test additions and land that.

@kevans91, yeah, I can confirm that it has been fixed. 
I've added a few commits to the current PR to remove all C code changes and left only test files changes.

bsdimp commented:
Closed by hand, something with amiss with my scripts.
